### ISSUE TITLE
Putting table rows in other components

### ISSUE THREAD

**Issue by growlsworth** [glittershark/reactable#156]
Hey @glittershark,

I'd like to have my table rows exist in a different file/component from the table component. Something like this:

```JavaScript
var 
  reactable = require('reactable'),
  Table = reactable.Table, 
  Tr = reactable.Tr, 
  Td = reactabke.Td;

RowComponent = React.createClass({
  render: function(){
    return (
      <Tr>
        <Td column="Name">{this.props.row.name}</Td>
      </Tr>
    );
  }
});

TableComponent = React.createClass({
  renderRow: function(row){
    var key = "row-"+row.id;
    return (<RowComponent key={key} row={row}/>);
  },
  renderRows: function(){
    this.state.rows.map(this.renderRow);
  },
  render: function(){
    return (<Table>{this.renderRows()}</Table>);
  }
});
```

However it doesn't render properly. Any thoughts? I've tried manually specifying columns, but that didn't make a difference. The closest I've gotten is it showing three blank rows.

Thanks!

**Comment by glittershark** [comment]
So, yeah, we should definitely allow this. The problem, at least with the Tr, is that the table parses through its children components and only pulls data from `Tr`s - maybe we descend a level down for each component and check the type of that?

**Comment by growlsworth** [comment]
Thanks for taking a look. If I remove the reactable elements and go with the standard `<table>`, `<tr>`, and `<td>` elements, then the html is correct, even if React thinks there are more virtual elements. I think it makes sense for reactable to try to be a drop in replacement for  basic `<table>`, `<tr>`, and `<td>` elements. To achieve this, I think you're right. It's a reasonable assumption to go 1 level deeper to find the Tr when it's a custom component.

I see utility in #159 as well. When the `<Td>` is complex and you might want to render different components depending on the need. Say for example that you want the cells to be editable so clicking on one transforms that cell into a form that can update the data value. That would make sense to have different custom React components for a single `<Td>`.

Thanks for taking a look.

**Comment by alfredo-cp** [comment]
@growlsworth exactly, I'm rendering a table which will have different type of content (buttons, dropdowns, links, etc) and I was thinking into create a base cell component wich will ask what type of cell to render, and also will validate data, etc.

I solved it for now just passing the normal ```Tr``` and my custom component as its children.

```
<CPTr column="Due Date">
  <CPDueDate>{item.due_date}</CPDueDate>
</CPTr>
```

**Comment by growlsworth** [comment]
This is proving to be more difficult than I would like. The issue is that I cannot get children of external sub components. Will require more investigation.

**Comment by hellogerard** [comment]
+1 for this. My table rows are pretty complex, and it makes sense to put them in their own Component.

Great component, BTW.

**Comment by aidanlister** [comment]
I'm trying to replicate an Angular UI table that looks like so: http://share.aidanlister.com/cjzX and finding it somewhat impossible in Reactable.

@growlsworth's renderRow looks like a really elegant way of handling it.

How far up the ladder you want to go in terms of declarative vs configuration is probably another discussion ... it would be cool to specify the renderer as part of the column definition perhaps?
```
      {
        key: 'tags',
        label: 'Tags',
        renderer: this.renderTagCell,
      },
      {
        key: 'ref',
        label: 'Ref'
      },
```

**Comment by jgarcia99** [comment]
So has this issue been solved? I also have a Row component with props in the td's nested in my Table component & I get the same console errors @alfredo-cp got.

**Comment by growlsworth** [comment]
There's an implementation discussion happening at https://github.com/glittershark/reactable/pull/166.

**Comment by jgarcia99** [comment]
@alfredo-cp what was your temporary resolution? Need some help

**Comment by jgarcia99** [comment]
```javascript
var Row = React.createClass({

    propTypes: {
        order: React.PropTypes.object.isRequired
    },

    getDefaultProps: function() {
        return {
            order : null
        };
    },

    onClickOrder: function() {
        rspActions.selectOrder(this.props.order);
    },

    render: function() {
        if (this.props.order) {
            return (
                <tr className="order-row" onClick={this.onClickOrder} style={this.styles.order}>
                    <td>{this.props.order.timeCreated}</td>
                    <td>{this.props.order.member.firstName} {this.props.order.member.lastName}</td>
                    <td>{this.props.order.vehicle.year} {this.props.order.vehicle.make} {this.props.order.vehicle.model} ({this.props.order.vehicle.color})</td>
                    <td>{this.props.order.orderID}</td>
                    <td>{this.props.order.type}</td>
                    <td><Status status={this.props.order.status} /></td>
                </tr>
            );
        }
    }

var Table = React.createClass({
    
    propTypes: {
        orders: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
    },
    
    getDefaultProps: function() {
        return {
            orders : []
        };
    },

    render: function() {

        var orders = '';
		if (this.props.orders) {
			orders = this.props.orders.map(function(order, index){
		      return <Row order={order} />;
		    });
		}

        return (
            <div>
                <Controls />
                <Bootstrap.Table>
                    <thead style={this.styles.thead}>
                        <tr>
                            <th>Date/Time</th>
                            <th>Name</th>
                            <th>Vehicle</th>
                            <th>Order Number</th>
                            <th>Type</th>
                            <th>Status</th>
                        </tr>
                    </thead>
                    <tbody style={this.styles.tbody}>
                        {orders}
                    </tbody>
                </Bootstrap.Table>
            </div>
        );
    },
```

**Comment by alfredo-cp** [comment]
@jgarcia99 I posted above what my temp solution was. I ended up doing my own grid component.

**Comment by jgarcia99** [comment]
@alfredo-cp you think I should go that route I see yours was td instead

**Comment by alfredo-cp** [comment]
@jgarcia99 if your data is complex rather than a plain object with strings I'd say go for your own grid component.

**Comment by jgarcia99** [comment]
Mine is just an array of objects. @alfredo-cp

**Comment by gravitycode** [comment]
+1 for this. I need split my complex rows implementation in other component

**Comment by dngster** [comment]
I have my table rows in separate components in standard html <table> <th> <tr> <td> similar to http://javaguirre.me/2014/02/11/twitter-streaming-api-with-node-socket-io-and-reactjs/ 
and when I'm stuck with #159 <td> errors. 
have also tried @alfredo-cp 's way of using a custom component inside <Tr> with no luck.
are there any other workaround solutions?

**Comment by muhammadahmed94** [comment]
Hi , I am using check boxes and check all. What should do to mark all my rows check on check all click without re rendering the grid? and what should I do forcefully re render grid or invoke **fetchdata** method on check box checked on unchecked?
